Enable to specify driver id by user.

jovan-wong opened the pull request:
## What do these changes do?
It's very useful for us to specify driver id to a driver.
In Java, we could specify it by `ray.conf`, and in Python, we should do it by `def init()`:
```python
ray.init(driver=b"01234567890123456789")
```
If we don't specify one, it will generate a random for us.

## Related issue number
N/A

robertnishihara commented on python/ray/worker.py:
```diff
@@ -1664,6 +1667,7 @@ def init(redis_address=None,
          plasma_directory=None,
          huge_pages=False,
          include_webui=True,
+         worker_id=None,
```
If this is the `driver_id` then it needs to be called `driver_i`.

robertnishihara commented on python/ray/worker.py:
```diff
@@ -2017,7 +2026,10 @@ def connect(info,
     assert worker.cached_functions_to_run is not None, error_message
     # Initialize some fields.
-    worker.worker_id = random_string()
+    if worker_id is None:
```
We should do some type checking. I think we want to assert that it is a Ray ObjectID or that it is 20 bytes or something like that. All of that is a bit ugly unfortunately.

robertnishihara commented:
I have mixed feelings about this. I think we want to restrict driver_id's to have type `ray.DriverID`, but we don't have that, so it would need to be `ray.ObjectID`, which feels weird because it's a driver, not an object. You're code example seems to use a short string, which doesn't make sense to me. Doesn't it need to be 20 bytes?

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Ray-PRB/8729/
Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Ray-PRB/8732/
Test PASSed.

jovan-wong commented:
@robertnishihara  Thanks a lot for your comments about it.

I have addressed some comments: 
adding type checking and renaming `worker_id` to `driver_id`.

Of course, it's too awkward for me to do type checking for these codes. As you mentioned, which type should `driver_id` should be? Maybe we should add an `UniqueID` type.

Another question I found is `worker.worker_id`'s type. I guess we assume that it's a string or bytes. One problem that comes with this is :
```python
    # Initialize some fields.
    if mode is WORKER_MODE:
        worker.worker_id = random_string()
    else:
        # This is the code path of driver mode.
        if driver_id is None:
            driver_id = ray.ObjectID(random_string())

        elif isinstance(driver_id, (bytes, str)):
            if len(driver_id) is not ray_constants.ID_SIZE:
                raise Exception("The length of given driver id str must be 20.")
            driver_id = ray.ObjectID(driver_id)

        elif not isinstance(driver_id, ray.ObjectID):
            raise Exception("The type of given driver id must be str or ray.ObjectID.")

        worker.worker_id = driver_id.id()
```
We should assume `driver_id` is a `ray.ObjectID` because we should cast from a str or a `ray.ObjectID`, And the next we should cast `driver_id` to str(or bytes) to match `worker.worker_id`.

I think the better way to solve it is to use `worker_id` as `ray.ObjectID` instead of `str(or bytes)`. And we needn't worry about the length of `worker_id` any more, right?

jovan-wong commented:
@guoyuhong  cc

robertnishihara commented:
@guoyuhong @jovan-wong what is the use case for specifying the driver ID?

raulchen commented:
Another purpose is that we want to have an external 'job manager' that runs all drivers with pre-defined driver ids, so it can manage the drivers by the ids.

@robertnishihara do you have any other questions? This PR looks good to me.

raulchen commented:
Another purpose is that we want to have an external 'job manager' that runs all drivers with pre-defined driver ids, so it can manage the drivers by the ids.

@robertnishihara do you have any other questions? This PR looks good to me.

raulchen commented:
Another purpose is that we want to have an external 'job manager' that runs all drivers with pre-defined driver ids, so it can manage the drivers by the ids.

@robertnishihara do you have any other questions? This PR looks good to me.

raulchen commented:
Another purpose is that we want to have an external 'job manager' that runs all drivers with pre-defined driver ids, so it can manage the drivers by the ids.

@robertnishihara do you have any other questions? This PR looks good to me.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Ray-PRB/8795/
Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Ray-PRB/8797/
Test FAILed.

robertnishihara commented on python/ray/worker.py:
```diff
@@ -2016,8 +2026,23 @@ def connect(info,
+        if driver_id is None:
+            driver_id = ray.ObjectID(random_string())
+
+        elif isinstance(driver_id, (bytes, str)):
```
Let's check that it is `bytes` for Python 2 and `str` for Python 3.

robertnishihara replied:
actually, for now, let's require it to be an `ObjectID`. I think the more restrictive we make it the better since it makes it easier to change later.

robertnishihara commented on python/ray/worker.py:
```diff
@@ -1731,6 +1736,7 @@ def init(redis_address=None,
             Store with hugetlbfs support. Requires plasma_directory.
         include_webui: Boolean flag indicating whether to start the web
             UI, which is a Jupyter notebook.
+        driver_id: The id of driver, and it will be ignored in non-driver mode.
```
id -> ID

remove the stuff after the `,`, there is no such thing as "non-driver" mode.

robertnishihara replied:
at least from the perspective of the user and `ray.init`, there is no such thing as "non-driver" mode.

jovan-wong commented:
@robertnishihara  Tks 4 ur comments.
I have addressed them, and could u plz c it again?

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Ray-PRB/9000/
Test FAILed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Ray-PRB/9001/
Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Ray-PRB/9002/
Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Ray-PRB/9014/
Test PASSed.
